src/main/cpp/src/cast_string_to_float.cu

```diff
@@ -0,0 +1,730 @@
+        __ballot_sync(0xffffffff, (warp_lane == 1 && (c == '-' || c == '+')));
+      auto const exp_sign =
+        exp_sign_mask ? __ballot_sync(0xffffffff, warp_lane == 1 && c == '-') ? -1 : 1 : 1;
+      auto const chars_to_skip = exp_sign_mask ? 2 : 1;
+      c                        = __shfl_down_sync(0xffffffff, c, chars_to_skip);
+      bpos += chars_to_skip;
+
+      // the largest valid exponent for a double is 4 digits (3 for floats).
+      int const num_chars = min(4, blen - bpos);
+
+      // handle any chars that are not actually digits
+      //
+      auto const non_digit_mask  = __ballot_sync(0xffffffff, warp_lane < num_chars && !is_digit(c));
+      auto const first_non_digit = __ffs(non_digit_mask);
+
+      int const num_digits = first_non_digit > 0 ? first_non_digit - 1 : num_chars;
+
+      if (num_digits == 0) {
+        valid  = false;
+        except = true;
+        return 0;
+      }
+
+      uint64_t const digit = warp_lane < num_digits
+                               ? static_cast<uint64_t>(c - '0') * ipow[(num_digits - warp_lane) - 1]
+                               : 0;
+      manual_exp           = WarpReduce(temp_storage).Sum(digit, num_digits) * exp_sign;
+      c                    = __shfl_down_sync(0xffffffff, c, num_digits);
+      bpos += num_digits;
+    }
+
+    return manual_exp;
+  }
+
+  __device__ void check_trailing_bytes()
+  {
+    if (blen - bpos > 0) {
+      // strip trailing f if it exists
+      // f is a valid character at the end of a float string
+      auto const f_mask = __ballot_sync(
+        0xffffffff, (warp_lane == 0 && (c == 'F' || c == 'f' || c == 'd' || c == 'D')));
+      if (f_mask > 0) {
+        c = __shfl_down_sync(0xffffffff, c, 1);
+        bpos++;
+      }
+    }
+
+    // nothing trailing
+    if (blen - bpos == 0) { return; }
+
+    // strip any whitespace
+    remove_leading_whitespace();
+
+    // invalid characters in string
+    if (blen - bpos > 0) {
+      valid  = false;
+      except = true;
+    }
+  }
+
+  // sets validity bits, updates outgoing validity count for the block and potentially sets the
+  // outgoing ansi_except field
+  __device__ void compute_validity(bool const valid, bool const except = false)
+  {
+    // compute null count for the block. each warp processes one string, so lane 0
+    // from each warp contributes 1 bit of validity
+    size_type const block_valid_count =
+      cudf::detail::single_lane_block_sum_reduce<block_size, 0>(valid ? 1 : 0);
+    // 0th thread in each block updates the validity count and (optionally) the ansi_except flag
+    if (threadIdx.x == 0) {
+      atomicAdd(valid_count, block_valid_count);
+
+      if (ansi_except && except) { atomicMax(ansi_except, num_rows - row); }
+    }
+
+    // 0th thread in each warp updates the validity
+    size_type const row_id = warp_id;
+    if (threadIdx.x % 32 == 0 && valid) {
+      // uses atomics
+      cudf::set_bit(validity, row_id);
+    }
+  }
+
+  T* out;
```

Comment:
Since these members are not public, please add prefix (`_`) so they can be recognized in the code so we can distinguish between member vars and local vars.